content/en/docs/concepts/cluster-administration/networking.md

```diff
@@ -32,6 +34,19 @@ different approach.
 
 To learn about the Kubernetes networking model, see [here](/docs/concepts/services-networking/).
 
+## Cluster Networking Types
+
+Kubernetes clusters, attending to the IP families configured, can be categorized into:
+
+- IPv4 only: Pod, Services and Node objects only contain IPv4 addresses
+- IPv6 only: Pod, Services and Node objects only contain IPv6 addresses
+- IPv4/IPv6 or IPv6/IPv4 dual-stack, may contain IPv4 and/or IPv6 addresses.
```

Comment:
So what is this addition for? Because it's not actually true; the network plugin's configuration determines whether pods are ipv4/ipv6/dual, kube-apiserver's confiugration determines whether services are ipv4/ipv6/dual, and the cloud provider's or kubelet's configuration decides whether nodes are ipv4/ipv6 or dual. Nothing requires them all to be configured the same way...

Reply:
we have e2e tests that for the order of the IPs and the preferred IP family to verify that clusters meet our definition of network model

Reply:
> same as nothing requires pods and nodes to be routable, is our definition of network model what requires them to be the same

Historically, we have jumped through hoops to _not_ require that... AFAIK everything works "as expected" when you have inconsistent IP-stack-ness. (If you have dual-stack pods and single-stack IPv4 services, then you just can't create services that point at IPv6 pod IPs. If you have single-stack IPv4 pods and dual-stack services, then your dual-stack services will just never have any IPv6 endpoints. If you have dual-stack pods and dual-stack services and single-stack nodes, then you can have dual-stack pod-network services but not dual-stack host-network services. Etc.)

I'd certainly be happy to say that kubernetes doesn't support clusters with inconsistent IP-stack-ness, but I feel like that's a _change_ in our position, not a clarification of it.

Reply:
> I'd certainly be happy to say that kubernetes doesn't support clusters with inconsistent IP-stack-ness

Actually, I wouldn't, since that would imply you can't live-migrate a cluster from single-stack to dual-stack.

Reply:
>  AFAIK everything works "as expected" when you have inconsistent IP-stack-ness

I disagree, the cluster will not be functional, that the network works or that the system allows you to do certain operation does not mean the cluster works correctly: controllers,  Incluser authentication , webhooks, DNS, and critical functionalities will not work  ... the network in kubernetes is more than the IPs , we are responsible of anything in the cluster that uses the network

> but I feel like that's a change in our position

is not a change in our position, we have e2e tests asserting this https://github.com/kubernetes/kubernetes/blob/master/test/e2e/network/dual_stack.go , once we define an e2e we define a behavior

Reply:
basically, a cluster with the wrong configuration will not even pass the conformance tests, means is not supported

Reply:
is that fair?

Reply:
> the cluster will not be functional

What, precisely, would not function? Yes, you can't have a dual-stack service pointing to a single-stack pod, but you can't have a dual-stack service pointing to a single-stack external host either. That's not "not functioning", that's just "you asked it to do something that doesn't make sense".

> we have e2e tests asserting this

We have e2e tests asserting that if you have a fully-dual-stack cluster, that things work in a certain way. I will grant you that we do not have e2e tests for partially-dual-stack clusters.

> a cluster with the wrong configuration will not even pass the conformance tests

which _conformance_ tests will it fail? Nothing in `test/e2e/network/dual_stack.go` is conformance.

Reply:
> since that would imply you can't live-migrate a cluster from single-stack to dual-stack.

temporary states are unrelated, this is an architectural choice not a hard rule, a cluster transitioning states, upgrading or migration does not apply

Reply:
Also, you cannot plausibly argue that we don't support running a single-stack IPv4 cluster on nodes that happen to also have IPv6 addresses.

Reply:
> which conformance tests will it fail? Nothing in test/e2e/network/dual_stack.go is conformance.

fair, all the e2e tests are part of the kubernetes definition, a subset of them are used by CNCF, but all those tests are defined by the SIGS and eventually can be promoted, dual stack is not mandatory so it will not promoted to CNCF conformance, but we added to establish a standardization and that distros and users and develops can have an homogeneus experience

> the cluster will not be functional

> If you have dual-stack pods and dual-stack services and single-stack nodes, then you can have dual-stack pod-network services but not dual-stack host-network services

ok, I overgeneralized, sorry, there are permutations that as you are pointing out will work, but do we want to support them? does it make sense? why we should not define the correct way to do it to avoid surprises or edge cases that work in some places and not in other? one of the success of kubernetes is that all flavors are "compatible"

With dual stack there are some combinations that will not work, per example if the apiserver primary family and the Node or Pod single stack family does not match, InCluster config will not work., or kubectl exec ...

Anyway, for me the important is the definition of IPv4 only and IPv6 only, If the dual stack point is contentious please let me know how to rephrase, or I just skip it and link to the dual stack existing docs https://kubernetes.io/docs/concepts/services-networking/dual-stack/

Reply:
ok, changed , PTAL , now there is no possible ambiguity,

Reply:
> Also, you cannot plausibly argue that we don't support running a single-stack IPv4 cluster on nodes that happen to also have IPv6 addresses.

This is user facing documentation, not developer facint, I just try to make this simple to understand because when I talk with users and customers a lot of them are confused about this topic,  and only power users or developers will create complex combinations of networking parameters, most people use whatever Openshift, GKE, EKS, AKS, ranchers provides ...

I don't think is worth our time to keep discussing this, if you prefer to not merge this I'm ok with that too, or if you think some part has to be reworded please suggest it

Reply:
> it's not actually true; the network plugin's configuration determines whether pods are ipv4/ipv6/dual, kube-apiserver's confiugration determines whether services are ipv4/ipv6/dual, and the cloud provider's or kubelet's configuration decides whether nodes are ipv4/ipv6 or dual. Nothing requires them all to be configured the same way...

Let's be wary of merging anything that can mislead. On the other hand, if Kubernetes can be set up in a way that doesn't pass conformance tests, and the proposed docs are correct for all clusters that pass our tests, I think they're fine to merge.

Which applies here?

Reply:
> ok, changed , PTAL , now there is no possible ambiguity,

/lgtm

> Let's be wary of merging anything that can mislead.

The latest version doesn't claim that "all components have to be in sync" like the original did. It just doesn't discuss the case where they don't, which I think is fine.

Reply:
> if Kubernetes can be set up in a way that doesn't pass conformance tests,

[I overgeneralized and Dan was right there that there are some combinations that may work ](https://github.com/kubernetes/website/pull/43964#discussion_r1396407933 ), at least I can't proof  which of all the permutations in a dual stack cluster will work, 

I know for sure that there are some combinations that can make conformance test to not pass because while working on dual stack implementations in kubeadm, kind and other installers I hit bugs on `kubectl exec` , webhooks, admission controllers that depends on the Node, Pods, Services objects IPs for the discovery and if they don't work if the order does not match in a dual stack clusters, and those functionalities are exercised in the Conformance tests.

But as I said,  I just wrote how dual-stack is implemented in most of the installers of Kubernetes I know of,